lib/BuildSystem/BuildSystem.cpp

```diff
@@ -2262,6 +2262,174 @@ class ArchiveTool : public Tool {
   }
 };
 
+#pragma mark - StaleFileRemovalTool implementation
+
+class StaleFileRemovalCommand : public Command {
+  std::string description;
+
+  std::vector<std::string> expectedOutputs;
+  std::vector<std::string> filesToDelete;
+
+  BuildValue priorValue;
+  bool hasPriorResult = false;
+
+  virtual void configureDescription(const ConfigureContext&, StringRef value) override {
+    description = value;
+  }
+
+  virtual void getShortDescription(SmallVectorImpl<char> &result) override {
+    llvm::raw_svector_ostream(result) << (description.empty() ? "Stale file removal" : description);
+  }
+
+  virtual void getVerboseDescription(SmallVectorImpl<char> &result) override {
+    getShortDescription(result);
+    llvm::raw_svector_ostream(result) << ", stale files: [";
+    for (auto fileToDelete : filesToDelete) {
```

Comment:
Yah, I was thinking of extracting this to a function, but then we'd compute this twice. Maybe extract and add a flag so that we can tell we did it.